singularity-eos/eos/eos_powermg.hpp

```diff
@@ -0,0 +1,440 @@
+    ind = ind + 1;
+    ak = ak * x / ind;
+    ek = pf * ak;
+    sum = sum + ak;
+  }
+  if (ind >= maxind) {
+#ifndef NDEBUG
+    PORTABLE_WARN("Part of Hugoniot Temperature not converged");
+#endif // NDEBUG
+  }
+  return sum;
+}
+
+PORTABLE_INLINE_FUNCTION Real PowerMG::AllHugPressureFromDensity(Real rho) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real value = 0.0;
+  if (eta <= 0.0) {
+    if (eta < _Pmin / _K0toK40[0]) {
+      value = _Pmin;
+    } else {
+      value = _K0toK40[0] * eta;
+    }
+  } else {
+    value = _HugPressureFromDensity(rho);
+  }
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::AllHugInternalEnergyFromDensity(Real rho) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  return _E0 + robust::ratio(eta, 2.0 * _rho0) * AllHugPressureFromDensity(rho);
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::AllHugTemperatureFromDensity(Real rho) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real value = 0.0;
+  if (eta <= 0.0) {
+    value = _T0 * exp(eta * _G0);
+  } else {
+    value = _HugTemperatureFromDensity(rho);
+  }
+  return value;
+}
+
+PORTABLE_INLINE_FUNCTION Real PowerMG::InternalEnergyFromDensityTemperature(
+    const Real rho, const Real temp, Real *lambda) const {
+  Real value = AllHugInternalEnergyFromDensity(rho) +
+               _Cv0 * (temp - AllHugTemperatureFromDensity(rho));
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::PressureFromDensityTemperature(
+    const Real rho, const Real temp, Real *lambda) const {
+  Real value = AllHugPressureFromDensity(rho) +
+               _G0 * _rho0 * _Cv0 * (temp - AllHugTemperatureFromDensity(rho));
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::EntropyFromDensityTemperature(const Real rho,
+                                                                     const Real temp,
+                                                                     Real *lambda) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real value = _S0 - _G0 * _Cv0 * eta + _Cv0 * std::log(temp / _T0);
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::TExpansionCoeffFromDensityTemperature(
+    const Real rho, const Real temp, Real *lambda) const {
+  Real value =
+      robust::ratio(_Cv0 * _rho0 * _G0, TBulkModulusFromDensityTemperature(rho, temp));
+  return value;
+}
+PORTABLE_INLINE_FUNCTION Real PowerMG::TBulkModulusFromDensityTemperature(
+    const Real rho, const Real temp, Real *lambda) const {
+  return BulkModulusFromDensityTemperature(rho, temp) -
+         _G0 * _G0 * _Cv0 * _rho0 * robust::ratio(_rho0, rho) * temp;
+}
+PORTABLE_INLINE_FUNCTION Real
+PowerMG::_compBulkModulusFromDensityTemperature(const Real rho, const Real temp) const {
+  Real eta = 1.0 - robust::ratio(_rho0, rho);
+  Real value = (1.0 - _G0 * eta / 2.0) * _K0toK40[0];
+  Real sum = 0.0;
+  for (int ind = _M; ind >= 1; ind--) {
+    sum = sum * eta + ind * _K0toK40[ind] * eta;
+  }
```

Comment:
It seems like this exact calculation is repeated a few times. Would there be benefit in making a helper function which computes this?

Reply:
I don't think so. Helper functions for something that fits on 3-4 lines I think is too much. But I am happy to hear your arguments.

Reply:
This is a personal preference thing. I tend to add helper functions for everything, as that way if I make an error (or want to tweak or debug something) I only have to change one place in the not, e.g., 5. 

However, it's a trade-off of many small things... lines of code added, amount the reader needs to jump around, etc. You don't need to change this if you like it as is.

Reply:
Done